File: unpack.py
import os
import json
import base64
import gzip
import re

html_path = "NEX EMS Redesign (Standalone).html"
output_dir = "unpacked"

if not os.path.exists(output_dir):
    os.makedirs(output_dir)

with open(html_path, "r", encoding="utf-8") as f:
    content = f.read()

# Extract manifest
manifest_match = re.search(r'<script type="__bundler/manifest">(.*?)</script>', content, re.DOTALL)
if not manifest_match:
    print("Could not find manifest!")
    exit(1)

manifest_json = json.loads(manifest_match.group(1).strip())

# Extract template
template_match = re.search(r'<script type="__bundler/template">(.*?)</script>', content, re.DOTALL)
if not template_match:
    print("Could not find template!")
    exit(1)

template_str = json.loads(template_match.group(1).strip())

# Extract ext_resources
ext_match = re.search(r'<script type="__bundler/ext_resources">(.*?)</script>', content, re.DOTALL)
ext_resources = json.loads(ext_match.group(1).strip()) if ext_match else []

# Save manifest files
print(f"Unpacking {len(manifest_json)} assets...")
for uuid, entry in manifest_json.items():
    data = base64.b64decode(entry["data"])
    if entry.get("compressed", False):
        try:
            data = gzip.decompress(data)
        except Exception as e:
            print(f"Failed to decompress {uuid}: {e}")
            
    # Try to determine file extension
    mime = entry["mime"]
    ext = "bin"
    if "javascript" in mime or "json" in mime:
        ext = "js"
    elif "css" in mime:
        ext = "css"
    elif "html" in mime:
        ext = "html"
    elif "png" in mime:
        ext = "png"
    elif "svg" in mime:
        ext = "svg"
    elif "woff2" in mime:
        ext = "woff2"
        
    filename = f"{uuid}.{ext}"
    filepath = os.path.join(output_dir, filename)
    
    with open(filepath, "wb") as out_f:
        out_f.write(data)
    print(f"Saved {filename} ({len(data)} bytes)")

# Save template.html
with open(os.path.join(output_dir, "template.html"), "w", encoding="utf-8") as out_f:
    out_f.write(template_str)

# Save metadata
metadata = {
    "ext_resources": ext_resources,
    "manifest_info": {uuid: {"mime": entry["mime"], "compressed": entry.get("compressed", False)} for uuid, entry in manifest_json.items()}
}
with open(os.path.join(output_dir, "metadata.json"), "w", encoding="utf-8") as out_f:
    json.dump(metadata, out_f, indent=2)

print("Unpacking complete!")

File: unpack.js
const fs = require('fs');
const path = require('path');
const zlib = require('zlib');

const htmlPath = path.resolve(__dirname, 'NEX EMS Redesign (Standalone).html');
const outputDir = path.resolve(__dirname, 'unpacked');

if (!fs.existsSync(outputDir)) {
  fs.mkdirSync(outputDir, { recursive: true });
}

console.log('Reading ' + htmlPath + '...');
const content = fs.readFileSync(htmlPath, 'utf8');

// Extract manifest
const manifestMatch = content.match(/<script type="__bundler\/manifest">(.*?)<\/script>/s);
if (!manifestMatch) {
  console.error('Could not find manifest!');
  process.exit(1);
}
const manifestJson = JSON.parse(manifestMatch[1].trim());

// Extract template
const templateMatch = content.match(/<script type="__bundler\/template">(.*?)<\/script>/s);
if (!templateMatch) {
  console.error('Could not find template!');
  process.exit(1);
}
const templateStr = JSON.parse(templateMatch[1].trim());

// Extract ext_resources
const extMatch = content.match(/<script type="__bundler\/ext_resources">(.*?)<\/script>/s);
const extResources = extMatch ? JSON.parse(extMatch[1].trim()) : [];

console.log('Unpacking ' + Object.keys(manifestJson).length + ' assets...');

for (const [uuid, entry] of Object.entries(manifestJson)) {
  let data = Buffer.from(entry.data, 'base64');
  if (entry.compressed) {
    try {
      data = zlib.gunzipSync(data);
    } catch (err) {
      console.error('Failed to decompress ' + uuid + ':', err);
    }
  }

  // Determine file extension
  const mime = entry.mime;
  let ext = 'bin';
  if (mime.includes('javascript') || mime.includes('json')) {
    ext = 'js';
  } else if (mime.includes('css')) {
    ext = 'css';
  } else if (mime.includes('html')) {
    ext = 'html';
  } else if (mime.includes('png')) {
    ext = 'png';
  } else if (mime.includes('svg')) {
    ext = 'svg';
  } else if (mime.includes('woff2')) {
    ext = 'woff2';
  }

  const filename = `${uuid}.${ext}`;
  const filepath = path.join(outputDir, filename);

  fs.writeFileSync(filepath, data);
  console.log('Saved ' + filename + ' (' + data.length + ' bytes)');
}

fs.writeFileSync(path.join(outputDir, 'template.html'), templateStr, 'utf8');

const metadata = {
  ext_resources: extResources,
  manifest_info: {}
};
for (const [uuid, entry] of Object.entries(manifestJson)) {
  metadata.manifest_info[uuid] = {
    mime: entry.mime,
    compressed: !!entry.compressed
  };
}
fs.writeFileSync(path.join(outputDir, 'metadata.json'), JSON.stringify(metadata, null, 2), 'utf8');

console.log('Unpacking complete!');
